pansat/download/accounts.py

```diff
@@ -2,54 +2,192 @@
 
 ###############################################################################
 # Handling identities
 ###############################################################################
 
 
+def get_password(check=False):
+    """
+    Check if password is provided as ``PANSAT_PASSWORD`` variable.
+    If this is not the case query user to input password.
+
+    Params:
+        check(``bool``): Whether or user should insert the password twice
+            to avoid spelling errors.
+
+    Return:
+        ``str`` containing the password
+    """
+    try:
+        password = os.environ["PANSAT_PASSWORD"]
+    except KeyError:
+        print("Please enter your pansat user password:")
+        password = getpass.getpass()
+
+        if check:
+            print("Please repeat the password:")
+            password_re = getpass.getpass()
+            if password != password_re:
+                print("The two passwords don't match!")
+                get_password(check=check)
+
+    return password
+
+
+def hash_password(password, salt):
+    """
+    Computes the hash of a password.
+
+    Args:
+        password(``str``): The password
+        slat(``bytes``): The salt to use to compute the hash.
+
+    Return:
+        The base64 encoded hash of the password.
+    """
+    password = password.encode()
+    kdf = PBKDF2HMAC(algorithm=hashes.SHA256(), length=32, salt=salt, iterations=100000)
+    return base64.urlsafe_b64encode(kdf.derive(password))
+
+
+def decrypt(secret):
+    """
+    Decrypt password.
+
+    Args:
+        secret(``str``): The encrypted password
+
+    Return:
+        The decrypted password.
+
+    Raises:
+
+        cryptography.fernet.InvalidToken: Raised if the encrypted password
+            is invalid.
+    """
+    fernet = Fernet(_PANSAT_SECRET)
+    return fernet.decrypt(secret).decode("utf8")
+
+
+def encrypt(password):
+    """
+    Encript password.
+
+    Args:
+        password(``str``): The password to encrypt.
+
+    Return:
+        The encrypted password as ``bytes``.
+    """
+    fernet = Fernet(_PANSAT_SECRET)
+    return fernet.encrypt(password.encode())
+
+
+def authenticate():
+    """
+    Verify pansat password.
+
+    Reads pansat password from environment variable ``PANSAT_PASSWORD``.
+    If this is not available, queries user to input password from standard
+    in. Verifies password using the hash stored in the identity list.
+
+    Raises:
+
+        Exception if the provided password is incorrect.
+    """
+    global _PANSAT_SECRET
+
+    if _PANSAT_SECRET is not None:
+        return
+
+    secret_hashed, salt = _IDENTITIES["pansat"]
+    secret_hashed = secret_hashed.encode()
+    salt = salt.encode()
+
+    password = get_password()
+
+    entered_secret = hash_password(password, salt)
+    entered_secret_hashed = hash_password(entered_secret.decode(), salt)
+
+    if secret_hashed != entered_secret_hashed:
+        raise Exception("The password you entered is incorrect.")
```

Comment:
a more specific exception? and also test it more specifically in test_parse_identity_failure